### ISSUE TITLE
Common background in batch processing

### ISSUE THREAD

**Issue by roaldarbol** [mooch443/trex#74]
Hi Tristan, hope you're well!

After massive struggles on the experimental side, I'm now finally beginning to do some actual tracking - finally! So now I'm back to provide - and get - some feedback. This is not a feature request really, more of a clarification question.

My experiments are quite long (24-48h), so I record videos in 6h blocks to be on the safe side. And so, I was wondering if there is a way to compute the background mean using all the videos to be used across all videos? (For my purpose, `max` has worked very well and seems very fast). 
I haven't done much batch processing, so along the same lines I struggled to use TGrabs (haven't actually gotten to the TRex part yet) for batch processing. Could you maybe share an example of how you achieve it in the terminal? E.g. process every file in a destination folder (specified with `-d`). That could maybe be worth adding to the `Batch Processing` docs page too I think. :-)

**Comment by mooch443** [comment]
Hey, hope you've arrived in 2022 alright! I am beginning to go through the messages I received over the break now, so hopefully I am not too late.

You can load an array of videos and terminate directly after generating the background image. This can be done by calling TGrabs with:

`tgrabs -i ["path/1","/path/2"] -quit_after_average -o output_name`

Hopefully this stuff still works, but this is the intended use. You can then choose any averaging_method that you prefer.

There is little documentation on batch processing unfortunately, that is true. :/ I should work on docs anyway. I always use some combination of bash scripting and tgrabs/trex stuff, like you could do:

```
ROOT=/videos
FILES=$(ls ${ROOT})
TGRABS_PARM="-averaging_method max -threshold 50 -use_closing"
TREX_PARM="-track_threshold 10 -track_max_individuals 10 -auto_quit"

for f in ${FILES}; do
    BASE=$(basename $f .mp4)
    CMD="tgrabs -d ${ROOT} -i ${f} ${TGRABS_PARM} -log_file convert_${BASE}.log"
    if eval $CMD; then
        echo "Converted ${BASE}."
        CMD="trex -d ${ROOT} -i ${BASE} ${TREX_PARM} -log_file track_${BASE}.log"
        if eval $CMD; then
            echo "Tracked ${BASE}"
        fi
    fi
done
```

Hope that helps! :)

**Comment by mooch443** [comment]
Oh, I might have misunderstood your message - you did not want to cut them together. If you just want to use the same background, simply rename it to have the same filename. This can be done in a bash script as well.

```
ROOT=/videos
FILES=$(ls ${ROOT})
TGRABS_PARM="-averaging_method max -threshold 50 -use_closing"
TREX_PARM="-track_threshold 10 -track_max_individuals 10 -auto_quit"

for f in ${FILES}; do
    BASE=$(basename $f .mp4)

    # copy bg to have the correct name:
    cp default.png ${ROOT}/average_${BASE}.png

    CMD="tgrabs -d ${ROOT} -i ${f} ${TGRABS_PARM} -log_file convert_${BASE}.log"
    if eval $CMD; then
        echo "Converted ${BASE}."
        CMD="trex -d ${ROOT} -i ${BASE} ${TREX_PARM} -log_file track_${BASE}.log"
        if eval $CMD; then
            echo "Tracked ${BASE}"
        fi
    fi
done
```

**Comment by roaldarbol** [comment]
Hey, and yes thanks had a good break - hope you're made it well into 2022 too! 
Yup, that makes sense. The idea was to have a background which was computed across all 24h and apply it to the 6h videos; but I made some changes in the setup  removing non-animal stuff which was moving a bit during the video (tissue paper). And now I'm ending up using 12h videos which works perfectly!

And thanks for the idea of using bash scripts, haven't worked much with bash other than the basic terminal fiddling, so this'll be great! I think it would be great to document especially this part as I imagine it'll be a very typical use case; fiddle around to find best parameters and then apply them across many videos.

A small (unrelated) question out of curiosity: I noticed that the `.pv` files are bigger than the original `.mp4` files, is this intended and why is that?

Thanks!

**Comment by roaldarbol** [comment]
So! Learned a bit of bash (try, fail, try again, and repeat), and came up with a neat solution. I modified the directory tree a bit, so now it looks like this:

```bash
├── experiments
│   ├── 2021-11-29
│   │   ├── environmental_data
│   │   ├── results
│   │   │   ├── ants-1.results.meta
│   │   │   ├── ants-1.settings
│   │   │   ├── ants-2.results.meta
│   │   │   ├── ants-2.settings
│   │   │   ├── data
│   │   │   │   ├── ants-1_beetle0.csv
│   │   │   │   ├── ants-1_beetle1.csv
│   │   │   │   ├── ants-1_beetle2.csv
│   │   │   │   ├── ants-1_beetle3.csv
│   │   │   │   ├── ants-1_beetle4.csv
│   │   │   │   ├── ants-1_beetle5.csv
│   │   │   │   ├── ants-2_beetle0.csv
│   │   │   │   ├── ants-2_beetle1.csv
│   │   │   │   ├── ants-2_beetle2.csv
│   │   │   │   ├── ants-2_beetle3.csv
│   │   │   │   ├── ants-2_beetle4.csv
│   │   │   │   └── ants-2_beetle5.csv
│   │   │   ├── track_ants-1.log
│   │   │   └── track_ants-2.log
│   │   ├── videos_pv
│   │   │   ├── ants-1.pv
│   │   │   ├── ants-2.pv
│   │   │   ├── average_ants-1.png
│   │   │   ├── average_ants-2.png
│   │   │   ├── convert_ants-1.log
│   │   │   └── convert_ants-2.log
│   │   └── videos_raw
│   │       ├── ants-1.mp4
│   │       └── ants-2.mp4
├── trex_batch.sh
└── trex_beetles.settings
```

The script (I call it `trex_batch.sh`) asks for the date of the experiment and a settings file, and is run like so (in case other bash-beginners come here) `/bin/bash trex_batch.sh`.

```bash
read -p 'Date of the experiment: ' DATE
read -p 'Settings file: ' SETTINGS

ROOT=$(pwd)
EWD=$ROOT/experiments/$DATE
PARAMS=$ROOT/$SETTINGS
FILES=$(find "$ROOT/experiments/$DATE/videos_raw" -type f -name *.mp4)

for f in ${FILES}; do
    echo $f
    BASE=$(basename $f .mp4)
    CMD="tgrabs \
        -d ${EWD}/videos_pv \
        -i ${f} \
        -s ${PARAMS} \
        -log_file convert_${BASE}.log"
    if eval $CMD; then
        echo "Converted ${EWD}."
        CMD="trex \
            -d ${EWD}/results \
            -i ../videos_pv/${BASE}.pv \
            -s ${PARAMS} \
            -log_file track_${BASE}.log \
            -auto_quit -auto_no_results"
        if eval $CMD; then
            echo "Tracked ${EWD}"
        fi
    fi
done
```

Feel free to give feedback if somethings comes to mind any/or use it for documentation if it's helpful. :-) Cheers!

**Comment by mooch443** [comment]
This looks great!

I suppose `bash trex_batch.sh` would suffice on most systems, but it doesnt hurt to make sure. Also, of course this find command finds all files recursively. So users have to make sure that they only have files in there that they want to have in there. It may also be safer to do `'*.mp4'` in the find command, to make sure that the * matching isnt executed until you want it to be.

It may be useful to grab also the error-out, which can be done by:

```bash
if { eval $CMD 2>&1; } > error.log; then
    echo "Success."
else
    echo "Failed."
    mv error.log "${EWD}_error.log"
fi

```

Not much more to add though, for now! I would probably let the date/time auto generate when I am recording, but this may vary depending on when you record/evaluate and such.

Thanks for posting this! I updated the docs as well at [https://trex.run/docs/batch.html](https://trex.run/docs/batch.html).
Have a look at the changes I made :-)
